sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/ContainerAPITest.groovy

```diff
@@ -1830,6 +1831,47 @@ class ContainerAPITest extends APISpec {
         thrown(BlobStorageException)
     }
 
+    def setupContainerForListing(BlobContainerClient containerClient) {
+        def blobNames = ["foo", "bar", "baz", "foo/foo", "foo/bar", "baz/foo", "baz/foo/bar", "baz/bar/foo"] as Set
+        def data = getRandomByteArray(Constants.KB)
+        def blobs = [] as Set
+
+        for (def blob : blobNames) {
+            def blockBlobClient = containerClient.getBlobClient(blob).getBlockBlobClient()
+            blobs.add(blockBlobClient)
+            blockBlobClient.upload(new ByteArrayInputStream(data), Constants.KB)
+        }
+    }
+
+    def "List blobs hier segment with version prefix and delimiter"() {
+        setup:
+        def options = new ListBlobsOptions()
+            .setDetails(new BlobListDetails().setRetrieveVersions(true))
+            .setPrefix("baz")
+
+        setupContainerForListing(cc)
+
+        def foundBlobs = [] as Set
+        def foundPrefixes = [] as Set
+
+        when:
+        cc.listBlobsByHierarchy("/", options, null).stream().collect(Collectors.toList())
+            .forEach { blobItem ->
+                if (blobItem.isPrefix()) {
+                    foundPrefixes << blobItem
+                } else {
+                    foundBlobs << blobItem
+                }
+            }
+
+        then:
+        foundBlobs.size() == 1
+        foundPrefixes.size() == 1
+        foundBlobs[0].getName() == "baz"
+        foundBlobs[0].getVersionId() != null
+        foundPrefixes[0].getName() == "baz/"
```

Comment:
Is this actually the expected outcome? Looking at the setup, blobs with prefix `baz` include `baz/foo`, `baz/foo/bar`, and `baz/bar/foo`. If I list under the prefix `baz`, I do not expect to find these results.

Reply:
This is the same as the results we got in the .NET SDK - https://github.com/Azure/azure-sdk-for-net/pull/30247/files#diff-cfbbd86d890e76d951dd6142c352ee932045fb0ce70ceb31c1d069f404c692c2